### PR TITLE
fix: strip trailing shell continuation backslashes

### LINKED ISSUE #254: [Medium] today: `docker ps\\` rendered with trailing backslash — un-stripped shell line continuation
## Summary
The `termstory today` command renders shell line continuation backslashes verbatim when they appear at the end of commands without actual continuations, displaying `docker ps\` instead of the intended `docker ps`. 

## Steps to Reproduce
1. Run a command in zsh with a trailing backslash but without continuing the line (e.g., `docker ps\` followed by Enter)
2. Run `PYTHONPATH=. python3 -m termstory today`
3. Observe that the command is rendered with the trailing backslash intact

## Expected vs Actual
**Expected**: Trailing backslashes that are shell line continuation markers without actual continuations should be stripped before display, showing `docker ps` instead of `docker ps\`. 

**Actual**: The trailing backslash is rendered verbatim in the output, making the report appear broken with commands like `docker ps\`.

## Environment
Not specified

## Context
The issue occurs in the `format_today_output` function in `termstory/formatter.py`, which processes commands for display using `clean_command_to_memory`.  The parser's `clean_command` function in `termstory/parser.py` handles multiline continuations when a backslash is followed by a newline, but does not strip trailing backslashes that lack continuations.

### PR DESCRIPTION
**Pull request by Ayush-0918**
## Summary

Fixes a bug where shell commands ending with an unescaped trailing backslash (e.g., `docker ps\`) were incorrectly displayed with the backslash even when no multiline continuation existed. The fix adds a final cleanup step in `clean_command()` that applies the shell even/odd backslash rule: a trailing backslash run of odd length is a line continuation (drop one), while an even-length run is a literal (keep). Literal backslashes inside paths (`ls C:\`) and unbalanced quotes (`echo "C:\`) are preserved.

### Core
- **termstory/parser.py**: Added `_quotes_balanced()` and `_strip_trailing_continuation()` helpers. The trailing-backslash cleanup in `clean_command()` now:
  - Only strips when quotes are balanced.
  - Applies the even/odd rule so `docker ps\` → `docker ps` but `ls C:\` and `echo \\` keep their literal backslashes.
  - Keeps the backslash when it is preceded by a path/quote character or sits inside an unbalanced quoted argument.

### Tests
- **tests/test_parser.py**: Added focused edge-case coverage in `test_clean_command_strips_only_unescaped_trailing_backslash()` verifying unescaped trailing backslashes are removed, escaped literal backslashes are preserved, even-length runs are kept, and a trailing backslash inside an unbalanced quoted argument or Windows path is retained.

## Fixes

- Fixes #254 — Commands ending with trailing backslash were incorrectly rendered with the backslash when no multiline continuation existed

## Breaking Changes

None

## Testing

```bash
python -m pytest tests/test_parser.py::test_clean_command tests/test_parser.py::test_clean_command_strips_only_unescaped_trailing_backslash -v
```

## Notes

The change is localized to the `clean_command()` function and only affects the final cleanup step after multiline commands have already been joined and whitespace normalized. Valid multiline continuations (e.g., `echo hello \\\n  world`) are still handled correctly while single-line commands with trailing backslashes are cleaned up without dropping literal backslashes.

### CODE REVIEW

**Comment by Ayush-0918** [comment]
:miku /ecsoc

**Review by bitflicker64 — CHANGES_REQUESTED** [review]
can u fix the suggestions from greptile

**Comment by Ayush-0918** [comment]
Thanks for the review! I reverted my previous approach after reviewing the Greptile feedback because it could remove legitimate trailing backslashes. I'm looking into a safer fix that only strips actual shell continuation markers without changing valid command text. Could you clarify the expected behavior for commands that intentionally end with a literal trailing backslash?

**Comment by bitflicker64** [comment]
Good instinct on not wanting to nuke legitimate backslashes — but the current `\\\s*$` in the diff is actually the thing the review flagged: it strips *any* trailing backslash, including an escaped literal.

The semantic that resolves your question: in shell, a **lone** trailing backslash is *always* a line-continuation marker. You can't have a "literal single trailing backslash" — it would escape the newline. A literal backslash is written doubled (`\\`). So the safe rule is: strip a trailing backslash **only when it is not preceded by another backslash**.

```python
cleaned = re.sub(r"(?<!\\)\\\s*$", "", cleaned)
```

I verified the behavior against the edge cases:

| input (history-stored) | current `\\\s*$` | proposed `(?<!\\)\\\s*$` |
|---|---|---|
| `docker ps\` (1 = continuation) | `docker ps` ✓ | `docker ps` ✓ |
| `echo \\` (2 = escaped literal) | `echo ` ✗ strips real text | `echo \\` ✓ keeps |
| `cd C:\\` (2 = escaped literal) | `cd C:` ✗ | `cd C:\\` ✓ keeps |
| `git status` | `git status` | `git status` |

So to answer your question directly: there is no "intentionally literal single trailing backslash" case to preserve — if it's a single `\`, it's a continuation and should be stripped; if it's a double `\\`, it's a literal and must stay.

Could you swap in the lookbehind version and add a test covering both the `docker ps\` → `docker ps` and `echo \\` → `echo \\` cases? That locks the behavior and closes #254 without the regression Greptile caught.

File: termstory/parser.py [hunk 1]
```diff
@@ -24,6 +24,8 @@ def clean_command(cmd_str: str) -> Optional[str]:
     
     cleaned = re.sub(r'\\\s*\n', ' ', cleaned)
     cleaned = " ".join(cleaned.split())
+    # Strip a trailing shell line-continuation backslash when it has no continuation.
+    cleaned = re.sub(r'\\\s*$', '', cleaned)
```
Comment (bitflicker64) [inline termstory/parser.py]: Following up with the exact swap as an inline suggestion so it's a one-click commit.

```suggestion
    cleaned = re.sub(r'(?<!\\)\\\s*$', '', cleaned)
```

Replaces `\\\s*$` (strips ANY trailing backslash) with `(?<!\\)\\\s*$` (only strips a continuation marker, never an escaped literal). Pair with the two test cases from my earlier comment and #254 closes without the regression Greptile flagged.

**Comment by Ayush-0918** [comment]
Addressed the requested review changes.

Updated the cleanup logic to strip only a lone trailing shell continuation backslash.
Added regression tests covering both:
removal of a trailing shell continuation (docker ps\ → docker ps)
preservation of escaped literal trailing backslashes.

All tests are passing locally. Please let me know if you'd like any further changes.

File: termstory/parser.py [hunk 2]
```diff
@@ -24,6 +40,12 @@ def clean_command(cmd_str: str) -> Optional[str]:
     
     cleaned = re.sub(r'\\\s*\n', ' ', cleaned)
     cleaned = " ".join(cleaned.split())
+    # Strip a trailing shell line-continuation backslash only when quotes are
+    # balanced. If a command reaches this point with unbalanced quotes, a
+    # remaining terminal backslash is part of a literal argument (e.g. echo "C:\)
+    # and must be preserved. Escaped backslashes (\\) are always preserved.
+    if _quotes_balanced(cleaned):
+        cleaned = re.sub(r"(?<!\\)\s*\\\s*$", "", cleaned)
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply termstory/parser.py]: fuck u

[the remark this answers is not in the document]
Comment (bitflicker64) [reply termstory/parser.py]: oh wait i got it

File: termstory/parser.py [hunk 3]
```diff
@@ -14,6 +14,52 @@
...
+    # If the backslash run is preceded by a path/quote character, it is almost
+    # certainly a literal (e.g. `ls C:\`, `cd "foo\`) rather than a continuation
+    # marker, so keep it. Otherwise treat an odd-length run as a line continuation
+    # and drop exactly one backslash.
+    if k < n and s[n - 1 - k] in (":", '"', "'", "/", "\\"):
+        return s
+    if k % 2 == 1:
+        # Drop only the continuation marker (and any trailing whitespace that
+        # preceded it), keep the literal pairs.
+        return s[: n - 1].rstrip()
```
[the remark this answers is not in the document]
Comment (bitflicker64) [reply termstory/parser.py:60]: Hi Greptile — thanks for the close read. I want to push back on this P1, because resolving it as suggested would re-introduce the exact bug this PR fixes (#254).

**Why the current behavior is correct:**

In POSIX shell, a trailing backslash *always* means "line continuation" — there is no syntactic way for a lone unescaped trailing backslash to be a literal. A literal trailing backslash must be doubled (`\\`). So:

- `docker ps\`  → continuation marker → strip → `docker ps`  ✅ (matches #254 expected output)
- `echo foo\`   → continuation marker → strip → `echo foo`  ✅ (same rule)
- `ls C:\`      → backslash preceded by `:` (path char) → keep → `ls C:\`  ✅ (literal Windows path)
- `echo \\`     → escaped pair → keep → `echo \`  ✅ (literal)
- `echo "C:\`   → unbalanced quote → keep → `echo "C:\`  ✅ (literal inside quote)

Your example `clean_command("echo foo\")` → `echo foo` is behaving exactly per shell semantics: the user typed `echo foo\` and pressed Enter, which *is* a line continuation. The command that actually ran is `echo foo`. So the stored/displayed command matching `echo foo` is correct, not a corruption.

**The only genuinely ambiguous case** is an unbalanced quoted argument like `echo "C:\`, where the backslash is a literal inside an unterminated quote. That case is now explicitly preserved (the `_quotes_balanced` guard + path/quote-char check).

So I've kept the strip for the clear continuation case and protected the literal cases. Reverting it would bring back `docker ps\` rendering verbatim, which is the reported defect. Happy to discuss if you see a case I'm missing.

**Review by bitflicker64 — APPROVED** [review]

**Comment by bitflicker64** [comment]
:miku /applaud @Ayush-0918
